tests/backup/backup_scale_test.go

```diff
@@ -0,0 +1,211 @@
+					backupLocationUIDMap[i] = uuid.New()
+					err := CreateBackupLocation(provider, backupLocationNameMap[i], backupLocationUIDMap[i], cloudCredName, cloudCredUID, getGlobalBucketName(provider), orgID, "")
+					dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of backup location [%s]", backupLocationNameMap[i]))
+				}
+			}
+		})
+		Step("Registering cluster for backup from px-admin", func() {
+			log.InfoD("Registering cluster for backup from px-admin")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			err = CreateApplicationClusters(orgID, "", "", ctx)
+			dash.VerifyFatal(err, nil, "Creating source and destination cluster")
+			log.InfoD("Verifying cluster status for both source and destination clusters")
+			clusterStatus, err := Inst().Backup.GetClusterStatus(orgID, SourceClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", SourceClusterName))
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", SourceClusterName))
+			clusterStatus, err = Inst().Backup.GetClusterStatus(orgID, destinationClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", destinationClusterName))
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", destinationClusterName))
+			clusterUid, err = Inst().Backup.GetClusterUID(ctx, orgID, SourceClusterName)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching [%s] cluster uid", SourceClusterName))
+		})
+		Step(fmt.Sprintf("Taking [%d] backup for the each application from px-admin", numberOfBackups), func() {
+			log.InfoD(fmt.Sprintf("Taking [%d] backup for the each application from px-admin", numberOfBackups))
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "failed to fetch ctx for admin")
+			createBackup := func(backupName string, namespace string, index int) {
+				defer GinkgoRecover()
+				defer wg.Done()
+				err := CreateBackup(backupName, SourceClusterName, backupLocationNameMap[index], backupLocationUIDMap[index], []string{namespace}, labelSelectors, orgID, clusterUid, "", "", "", "", ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation and validation of backup [%s] of namespace (scheduled Context) [%s]", backupName, namespace))
+			}
+			for _, namespace := range bkpNamespaces {
+				for index := 0; index < numberOfBackups; index++ {
+					time.Sleep(timeBetweenConsecutiveBackups)
+					backupName := fmt.Sprintf("%s-%s-%s", BackupNamePrefix, backupLocationNameMap[index], RandomString(4))
+					userBackupMap[index] = make(map[string]string)
+					userBackupMap[index][backupName] = namespace
+					wg.Add(1)
+					go createBackup(backupName, namespace, index)
+					// wait for max concurrent backup limit of 4 to be completed
+					if index%4 == 0 {
+						wg.Wait()
+					}
+				}
+			}
+			wg.Wait()
+		})
```

Comment:
How about making using of semaphore?

```
semaphore := make(chan int, 4)
for _, namespace := range bkpNamespaces {
	for index := 0; index < numberOfBackups; index++ {
		time.Sleep(timeBetweenConsecutiveBackups)
		backupName := fmt.Sprintf("%s-%s-%s", BackupNamePrefix, backupLocationNameMap[index], RandomString(4))
		userBackupMap[index] = make(map[string]string)
		userBackupMap[index][backupName] = namespace
		wg.Add(1)
		semaphore <- 0
		go func(backupName string, namespace string, index int) {
			defer wg.Done()
			defer func() {
				<-semaphore
			}()
			createBackup(backupName, namespace, index)
		}(backupName, namespace, index)
	}
}
wg.Wait()
```

I have not tested it.

Reply:
Added  semaphore and it working as expected, Thanks for the suggestion @kphalgun-px 
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/2988/